src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -0,0 +1,756 @@
+          currPos = nextPos;
+        } else {
+          // Reduce first bigger node on level by one
+          // and descend to next level
+          next.setInBetweenCount(next.getInBetweenCount(currLvl) - 1, currLvl);
+          currLvl--;
+        }
+      }
+    }
+
+    removeNode(currNode);
+    return currNode.getValue();
+  }
+
+  @Override
+  public int rankOf(Object pO) {
+    Preconditions.checkNotNull(pO);
+
+    @SuppressWarnings("unchecked")
+    T val = (T) pO;
+    @Var Node<T> currNode = head;
+    @Var Node<T> next;
+    @Var int index = -1;
+    for (int currLvl = MAX_LEVEL; currLvl >= LEVEL_ONE; currLvl--) {
+      next = currNode.getNext(currLvl);
+      while (next != null && comparator.compare(val, next.getValue()) >= 0) {
+        currNode = next;
+
+        int comp = comparator.compare(val, currNode.getValue());
+        if (comp >= 0) {
+          index += next.getInBetweenCount(currLvl);
+        }
+
+        if (comp == 0) {
+          break;
+        } else {
+          next = currNode.getNext(currLvl);
+        }
+      }
+    }
+
+    while (currNode != head && comparator.compare(val, currNode.getValue()) == 0) {
+      if (currNode.getValue().equals(pO)) {
+        return index;
+      } else {
+        currNode = currNode.getNext(LEVEL_ONE);
+        index++;
+      }
+    }
+
+    return -1;
+  }
+
+  @Override
+  public T first() {
+    if (size <= 0) {
+      throw new NoSuchElementException();
+    }
+    return getByRank(0);
+  }
+
+  @Override
+  public T last() {
+    if (size <= 0) {
+      throw new NoSuchElementException();
+    }
+    return tail.getValue();
+  }
+
+  @Override
+  public SortedSet<T> subSet(T pFromElement, T pToElement) {
+    Preconditions.checkNotNull(pFromElement);
+    Preconditions.checkNotNull(pToElement);
+    SkipList<T> subList = new SkipList<>(comparator);
+
+    int start = rankOf(pFromElement);
+    if (start < 0) {
+      throw new IllegalStateException("From-element doesn't exist in list: " + pFromElement);
+    }
+
+    int end = rankOf(pToElement);
+    if (end < 0) {
+      throw new IllegalStateException("To-element doesn't exist in list: " + pToElement);
+    }
+
+    Node<T> startNode = getNode(start);
+    Node<T> endNode = getNode(end);
+
+    for (Node<T> currNode = startNode; currNode != endNode; ) {
+      boolean existed = subList.add(currNode.getValue());
+      assert !existed;
+      currNode = currNode.getNext(LEVEL_ONE);
+    }
+
+    return subList;
+  }
+
+  @Override
+  public SortedSet<T> headSet(T pToElement) {
+    Preconditions.checkNotNull(pToElement);
+    return subSet(first(), pToElement);
+  }
+
+  @Override
+  public SortedSet<T> tailSet(T pFromElement) {
+    Preconditions.checkNotNull(pFromElement);
+    return subSet(pFromElement, last());
+  }
+
+  @Override
+  public Comparator<? super T> comparator() {
+    return comparator;
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return new ListIterator<T>() {
```

Comment:
Why implement `ListIterator` if the return type is just `Iterator` anyway?

We could change the return type to `ListIterator`, or expose an additional `listIterator` method like `List` has. But probably exposing a `ListIterator` is not a good idea because it would allow `add` and `set` at a given position.

Reply:
> We could change the return type to ListIterator, or expose an additional listIterator method like List has. But probably exposing a ListIterator is not a good idea because it would allow add and set at a given position.

Yes, I thought of both, mainly because I implemented the `List`-interface, at first,
but did choose not to do either because of the same reasons.

I just kept the implementation in case it is required at some point, but keeping code lying around is a bad practice, of course. I'll remove it.